src/server/qtquick/private/wquickxwayland.cpp

```diff
@@ -250,36 +250,11 @@ void WQuickXWayland::removeToplevel(WXWaylandSurface *surface)
 WXWaylandSurfaceItem::WXWaylandSurfaceItem(QQuickItem *parent)
     : WSurfaceItem(parent)
 {
-
-}
-
-WXWaylandSurfaceItem::~WXWaylandSurfaceItem()
-{
-
-}
-
-WXWaylandSurface *WXWaylandSurfaceItem::surface() const
-{
-    return m_surface;
-}
-
-void WXWaylandSurfaceItem::setSurface(WXWaylandSurface *surface)
-{
-    if (m_surface == surface)
-        return;
-
-    if (m_surface) {
-        bool ok = m_surface->disconnect(this);
-        Q_ASSERT(ok);
-    }
-
-    m_surface = surface;
-
-    if (surface) {
-        Q_ASSERT(surface->surface());
-        WSurfaceItem::setSurface(surface->surface());
-        connect(surface, &WXWaylandSurface::surfaceChanged, this, [this] {
-            WSurfaceItem::setSurface(m_surface->surface());
+    connect(this,&WXWaylandSurfaceItem::shellSurfaceChanged,[this]{
```

Comment:
Don't use signal here, can make `WSurfaceItem::setShellSurface` to virtual? for an example:

````
bool WSurfaceItem::setShellSurface(...) {
    if (m_shellSurface == newSurface) return false;
}
````

````
WXWaylandSurfaceItem::setShellSurface(...) override {
    if (WSurfaceItem::setShellSurface(...)) return false;
}
````